tests/generative/variable_strategies.py

```diff
@@ -219,3 +221,112 @@ def composite_filter(draw, wrapped, predicate):
     example = draw(wrapped)
     hyp.assume(predicate(example))
     return example
+
+
+def uses_the_database(v):
+    database_uses = [SelectPatientTable, SelectTable]
+    return any(t in database_uses for t in node_types(v))
+
+
+# SQlAlchemy does a certain amount of interpretation and optimization client-side. In particular in some cases where it
+# can statically determine that the result of a function is constant, it replaces that function with the constant
+# value. In extreme cases this may mean that the result of an entire query is completely determined client-side. In
+# those cases our engine cannot determine the population for the dataset because it draws the patient universe from all
+# tables used by the query -- and the query will use no tables.
+#
+# So we need to exclude such queries. This function detects function calls that SQLAlchemy will optimize away so that
+# can avoid queries made up entirely of such things.
+#
+# Note that it is not sufficient to check that there is at least one non-trivial node in the tree because the collapse
+# of trivial nodes lower down the tree can cause those higher up to become collapsible. For example this query reduces
+# to `False`:
+#
+#     Function.And(
+#         lhs=SelectColumn(
+#             source=SelectPatientTable(name="p0"),
+#             name="b1",
+#         ),
+#         rhs=Function.IsNull(source=Value(value=0)),
+#     )
+#
+# The only way to detect cases like that would be to replicate SQLAlchemy's interpretation of the entire tree in order
+# to check that the query is non-trivial.
+#
+# We therefore exclude all trivial nodes from the query. That means that there are realistic queries with trivial
+# subtrees that we can't generate with these strategies. We could consider composing known-non-trivial subtrees with
+# possibly-trivial subtrees as a future improvement to these strategies.
+#
+# And advantage to this approach is that we reject huge numbers of trivial queries that Hypothesis was otherwise
+# inclined to generate, so the queries generally are rather more realistic.
+def is_trivial(node):
+    if type(node) in [
+        Function.Not,
+        Function.IsNull,
+        Function.Negate,
+        Function.RoundToFirstOfMonth,
+        Function.RoundToFirstOfYear,
+        Function.YearFromDate,
+    ]:
```

Comment:
What happens when we add more unary functions?

Reply:
It's possible we could determine this (and the other lists of functions) dynamically by introspection over all node types exported by the query model i.e. walk over everything in `query_model.__all__`, is it a QueryNode, if so use `dataclasses.fields()` to check what it accepts.

Reply:
If we add a function that needs to be here then we'll get a test failure when we generate a query that get's unhelpfully optimized away by SQLAlchemy. It may be non-obvious that that's the problem, so a generic approach here would be a _bit_ better, but not worth the effort until we see an actual cost to the current approach, I think.